src/main/scala/io/iohk/ethereum/ledger/InMemoryWorldStateProxy.scala

```diff
@@ -0,0 +1,267 @@
+import io.iohk.ethereum.network.p2p.messages.PV63.AccountImplicits._
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.vm.{UInt256, Storage, WorldStateProxy}
+
+object InMemoryWorldStateProxy {
+
+  val byteArrayUInt256Serializer = new ByteArrayEncoder[UInt256] {
+    override def toBytes(input: UInt256): Array[Byte] = input.bytes.toArray[Byte]
+  }
+
+  val rlpUInt256Serializer = new RLPByteArraySerializable[UInt256]
+  implicit val accountSerializer = new RLPByteArraySerializable[Account]
+
+  def apply(
+    storages: BlockchainStorages,
+    stateStorage: NodeStorage,
+    stateRootHash: Option[ByteString] = None): InMemoryWorldStateProxy = {
+
+    val stateStorageProxy = InMemorySimpleMapProxy.wrap(stateStorage)
+    val accountsStateTrieProxy = createProxiedAccountsStateTrie(
+      stateStorageProxy,
+      stateRootHash.getOrElse(ByteString(MerklePatriciaTrie.calculateEmptyRootHash(kec256(_: Array[Byte]))))
+    )
+    val evmCodeStorageProxies = InMemorySimpleMapProxy.wrap(storages.evmCodeStorage)
+
+    val getBlockHashByNumber = (number: BigInt) => BlockchainImpl(storages).getBlockHeaderByNumber(number).map(_.hash)
+
+    new InMemoryWorldStateProxy(stateStorageProxy, accountsStateTrieProxy, Map.empty, evmCodeStorageProxies, Map.empty, getBlockHashByNumber)
+  }
+
+  def persistIfHashMatches(stateRootHash: ByteString, worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    val commited = commitState(worldState)
+    if(commited.stateRootHash == stateRootHash) persistStorages(commited)
+    else worldState
+  }
+
+  /**
+    * Updates state trie with current changes but does not persist them into the storages. To do so it:
+    *   - Commits code (to get account's code hashes)
+    *   - Commits constract storages (to get account's contract storage root)
+    *   - Updates state tree
+    *
+    * @param worldState Proxy to commit
+    * @return Updated world
+    */
+  private def commitState(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    def commitCode(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+      worldState.accountCodes.foldLeft(worldState) {
+        case (updatedWorldState, (address, code)) =>
+          val codeHash = kec256(code)
+          updatedWorldState.copyWith(
+            accountsStateTrie = updatedWorldState.accountsStateTrie +
+              (address.bytes -> updatedWorldState.getGuaranteedAccount(address).copy(codeHash = codeHash)),
+            evmCodeStorage = updatedWorldState.evmCodeStorage + (codeHash -> code),
+            accountCodes = Map.empty
+          )
+      }
+    }
+
+    def commitContractStorage(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy =
+      worldState.contractStorages.foldLeft(worldState) { case (updatedWorldState, (address, storageTrie)) =>
+        val persistedStorage = storageTrie.persist()
+        val newStorageRootHash = persistedStorage.inner.wrapped.getRootHash
```

Comment:
maybe we can add some abstraction around `persistedStorage.inner.wrapped.getRootHash` should we know inner structure of `persistedStorage` here?